x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsPersister.java

```diff
@@ -92,18 +106,18 @@ public Builder bulkPersisterBuilder(String jobId) {
     }
 
     public class Builder {
-        private BulkRequest bulkRequest;
+        private final Map<String, IndexRequest> items;
         private final String jobId;
         private final String indexName;
         private Supplier<Boolean> shouldRetry = () -> true;
 
         private Builder(String jobId) {
-            this.bulkRequest = new BulkRequest();
+            this.items = new LinkedHashMap<>();
             this.jobId = Objects.requireNonNull(jobId);
             this.indexName = AnomalyDetectorsIndex.resultsWriteAlias(jobId);
         }
 
-        public Builder shouldRetry(Supplier<Boolean> shouldRetry) {
+        public synchronized Builder shouldRetry(Supplier<Boolean> shouldRetry) {
```

Comment:
This isn't because of this PR but it might be a good chance to tidy up. `shouldRetry` seems to only be set once per instance of the `Builder` class. I would suggest making it a final field and requiring it in the constructor to get rid of this setter entirely.